packages/common/core-interfaces/src/jsonDeserialized.ts

```diff
@@ -0,0 +1,71 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import type { InternalUtilityTypes } from "./exposedUtilityTypes.js";
+
+/**
+ * Options for {@link JsonDeserialized}.
+ *
+ * @beta
+ */
+export interface JsonDeserializedOptions {
+	/**
+	 * A type that is managed by custom deserialization logic (beyond JSON.parse).
+	 *
+	 * The default value is `never`.
+	 */
+	Replaced: unknown;
+}
+
+/**
+ * Produces a type that results from a type `T` serialized and deserialized
+ * through JSON.
+ *
+ * @typeParam T - The type that was serialized.
+ *
+ * @remarks
+ * When used as a filter to inferred generic `T`, a compile-time error can be
+ * produced trying to assign `JsonDeserialized<T>` to `T`.
+ *
+ * Deserialized JSON never contains `undefined`, `symbol`, or function values.
+ * Object properties with values of those types are absent. So properties
+ * become optional (when there are other supported types in union) or are
+ * removed (when nothing else in union is supported).
+ * In an array, such values are replaced with `null`.
+ *
+ * `bigint` valued properties are simply removed as serialization attempts
+ * will throw.
```

Comment:
There are a lot of things this doc comment does not cover, despite it seeming to make an attempt to be complete, so I'll mention them here:

- Its perfectly possible to losslessly encode bigints into and from JSON. JSON allows arbitrary numeric precision. This doc comment makes no mention of JSON.strinify, so applying its limitation here seems arbitrary. There even is https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/rawJSON which makes it possible to to losslesslyencode bigints using JSON.stringify ( https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON#using_json_numbers )
- No mention of what subset of properties are considered at runtime. Typically, this is enumerable own properties. I thing the section on setters/getters is simply wrong. I'm not sure what you mean by "backed by a function": setters and getters always use functions. Assuming we are talking about JSON.strinify and similar functions what actually matters is if the property is an enumerable own property. Setter don't matter and getters are not special really as they are included when enumerable own. Since TypeScript does not capture ownness or enumerableness of properties in its type system, and most serializers including our stringify and toJson rely on them think this needs to be called out explicitly.
- No mention of the JSON limits on strings or numbers. Strings with unmatched utf16 surrogate pairs will not round trip, nor will NaN or infinities.
- You mention recursive types, but not recursive values

Generally the existing doc comment on Jsonable seems to cover a lot of details which are missing here (ex: the special number cases, assumptions about enumerability of properties, issues with ArrayLike, recursive object issues. If this is intending to replace that API, it seems like its docs should be at least as complete.

Reply:
Maybe this should link JsonSerializable's docs to cover all these cases, and document (at least in a private remarks) why we need both and how they differ?

Reply:
I've enhanced what it says about what it does without calling out an explicit comparison. And have added limitation notes. Please have another look.